### ISSUE TITLE
Resolve toolkit versioning

### ISSUE THREAD

**Issue by ddebrunner** [IBMStreams/administration#5]
Be good to get guidelines on how to version toolkits, otherwise it's likely that multiple different copies of the toolkit will exist with the same version number, as commits are made and people copy the code without using a "formal" release.

https://github.com/IBMStreams/toolkits/wiki/Toolkit-Versions

**Comment by chanskw** [comment]
I'd like to close on this.

I think it is a toolkit's committers team's decision on how to version and maintain their toolkits.  i.e. IBMStreams administration should not mandate how this has to be done.  However, I think having some general guidelines is useful.

I propose the following guidelines:

A toolkit version contains 3 digits.  Changes to each of the digits represent what are changed in the toolkit.  The assumption we make here is that user will specify dependencies like this in their applications:  [1.0.0, 2.0.0) - i.e. from 1.0.0 to 2.0.0 (exclusive)
1)  If the toolkit only contains minor fixes, bump the 3rd digit.
2)  If the toolkit contains significant new functions, but it will continue to be compatible with previous version, bumb the 2nd digit
3)  If the toolkit contains incompatible changes, bump the first digit.

I believe this is the minimum guideline for toolkit versioning.  Each toolkit repository should document their versioning strategy or simply refer to the wiki if they decide to follow this guideline.

e.g. If a repository wants to implement the qualifier approach, they are free to do so.  The minimum is that every toolkit release should require a version bump to identify what changes are made.

Please +1 on this proposal if you agree, and I can update the wiki.
Please -1 if you have any concern.  

I propose to leave voting for 1 week.  If there is no objection, I will also update the wiki.

Thanks!

**Comment by mikespicer** [comment]
+1 I would also be more explicit and add that 2 releases with different code should not have the same version number.

**Comment by chanskw** [comment]
Thanks Mike!  I agree... will add that.

**Comment by ddebrunner** [comment]
Minor nit: "A toolkit version contains 3 digit and a qualifier".

**Comment by ddebrunner** [comment]
> I think it is a toolkit's committers team's decision on how to version and maintain their toolkits
> IBMStreams administration should not mandate how this has to be done

I'm not so sure, any other scheme is likely to confuse users of the toolkit.

Also the project incubation page says: "Project ... follows semantic versioning (http://semver.org/)"
https://github.com/IBMStreams/administration/wiki/Project-Incubation-Process

What flexibility might a project have while still being a toolkit that can be used easily by others?

Are you proposing the incubation policy will change or that a toolkit can have a different scheme but won't be able to graduate?

**Comment by chanskw** [comment]
I think we are in agreement that toolkits should follow the semantic versioning guidelines.

So, let me rephrase this.

Here's what we are going to state:
It is the toolkit's committers team's decision on how to version and maintain their toolkits.  It is highly recommended that they follow the semantic versioning guidelines as documented.  Following this guideline is a requirement for a project to graduate from incubation.

My thoughts on this:
1)  This is a guideline that people need to learn and understand.  Prior to graduation, toolkit project may be more relaxed in how they manage their versions.  It may not be clear, but it is something the project team needs to learn and develop on.
2.  I think each project has different requirements and details on how to actually implement the versioning guidelines.  I do not think we can mandate the exact detail on how this is done, but we need to ensure that version numbers are bumped and reflect what's changed.  We also should make sure that the projects document how they have implemented the guideline.

Sound ok?

**Comment by chanskw** [comment]
I updated this page:
https://github.com/IBMStreams/administration/wiki/Toolkit-Versions

If you disagree, please update the page directly and we can discuss.  I hope we are close to closing this.

Thanks!

**Comment by chanskw** [comment]
Closing this one... please open new issue if we need more discussions.
